addons/payment_alipay/models/payment_transaction.py

```diff
@@ -0,0 +1,147 @@
+from odoo.tools.float_utils import float_compare
+
+from odoo.addons.payment_alipay.controllers.main import AlipayController
+
+_logger = logging.getLogger(__name__)
+
+
+class PaymentTransaction(models.Model):
+    _inherit = 'payment.transaction'
+
+    # === CRUD METHODS ===#
+
+    @api.model_create_multi
+    def create(self, vals):
+        res = super().create(vals)
+        for tx in res:
+            tx._check_alipay_configuration()
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('currency_id') or vals.get('acquirer_id'):
+            for tx in self:
+                tx._check_alipay_configuration()
+        return res
+
+    def _check_alipay_configuration(self):
+        if self.acquirer_id and self.acquirer_id.provider == 'alipay' and self.acquirer_id.alipay_payment_method == 'express_checkout':
+            if self.currency_id.name != 'CNY':
+                _logger.info("Only CNY currency is allowed for Alipay Express Checkout")
+                raise ValidationError("Alipay: " + _(
+                    "Only transactions in Chinese Yuan (CNY) are allowed for Express Checkout.\n"
+                    "If you wish to use another currency than CNY for your transactions, switch\n"
+                    "your configuration to a Cross-border account in Odoo."
+                ))
+
+    # === BUSINESS METHODS ===#
+
+    @api.model
+    def _get_tx_from_feedback_data(self, provider, data):
+        if provider != 'alipay':
+            return super()._get_tx_from_feedback_data(data, provider)
+        # verify signature
+        alipay_provider = self.env['payment.acquirer'].search([('provider', '=', 'alipay')], limit=1)
+        sign_check = alipay_provider._alipay_build_sign(data)
+        txn_id, sign = data.get('trade_no'), data.get('sign')
+        reference = data.get('reference') or data.get('out_trade_no')
+        if sign != sign_check:
+            _logger.info('Alipay: invalid sign, received %s, computed (%s), for data (%s)' % (sign, sign_check, data))
+            raise ValidationError('Alipay: ' + _('invalid sign: received %(sign)s, computed %(sc)s, for data %(data)s',
+                                                 sign=sign, sc=sign_check, data=data))
+
+        if not reference or not txn_id:
+            _logger.info('Alipay: received data with missing reference (%s) or txn_id (%s)' % (reference, txn_id))
+            raise ValidationError("Alipay: " +_('received data with missing reference %(r)s or txn_id %(t)s',
+                                                r=reference, t=txn_id))
+
+        tx = self.env['payment.transaction'].search([('reference', '=', reference)])
+        if not tx or len(tx) > 1:
+            raise ValidationError(
+                "Alipay: " + _(
+                    "received data with reference %(ref)s matching %(num_tx)d transaction(s)",
+                    ref=reference, num_tx=len(tx)
+                )
+            )
```

Comment:
My bad, there is no way that two or more could be matched here since there is a postgres constraint on the ref. (corrected in adyen too)
```suggestion
        if not tx:
            raise ValidationError(
                "Alipay: " + _("received data with reference %s matching no transaction")
            )
```